dace/transformation/interstate/state_fusion_with_happens_before.py

```diff
@@ -0,0 +1,589 @@
+            return False
+
+        return True
+
+    def apply(self, _, sdfg):
+        first_state: SDFGState = self.first_state
+        second_state: SDFGState = self.second_state
+
+        # Remove interstate edge(s)
+        edges = sdfg.edges_between(first_state, second_state)
+        for edge in edges:
+            if edge.data.assignments:
+                for src, dst, other_data in sdfg.in_edges(first_state):
+                    other_data.assignments.update(edge.data.assignments)
+            sdfg.remove_edge(edge)
+
+        # Special case 1: first state is empty
+        if first_state.is_empty():
+            sdutil.change_edge_dest(sdfg, first_state, second_state)
+            sdfg.remove_node(first_state)
+            if sdfg.start_state == first_state:
+                sdfg.start_state = sdfg.node_id(second_state)
+            return
+
+        # Special case 2: second state is empty
+        if second_state.is_empty():
+            sdutil.change_edge_src(sdfg, second_state, first_state)
+            sdutil.change_edge_dest(sdfg, second_state, first_state)
+            sdfg.remove_node(second_state)
+            if sdfg.start_state == second_state:
+                sdfg.start_state = sdfg.node_id(first_state)
+            return
+
+        # Normal case: both states are not empty
+
+        # Find source/sink (data) nodes
+        first_input = [node for node in first_state.source_nodes() if isinstance(node, nodes.AccessNode)]
+        first_output = [node for node in first_state.sink_nodes() if isinstance(node, nodes.AccessNode)]
+        second_input = [node for node in second_state.source_nodes() if isinstance(node, nodes.AccessNode)]
+
+        top2 = top_level_nodes(second_state)
+
+        # first input = first input - first output
+        first_input = [
+            node for node in first_input if next((x for x in first_output if x.data == node.data), None) is None
+        ]
+
+        # NOTE: We exclude Views from the process of merging common data nodes because it may lead to double edges.
+        second_mid = [
+            x for x in list(nx.topological_sort(second_state._nx)) if isinstance(x, nodes.AccessNode)
+            and second_state.out_degree(x) > 0 and not isinstance(sdfg.arrays[x.data], dt.View)
+        ]
+
+        # Merge second state to first state
+        # First keep a backup of the topological sorted order of the nodes
+        sdict = first_state.scope_dict()
+        order = [
+            x for x in reversed(list(nx.topological_sort(first_state._nx)))
+            if isinstance(x, nodes.AccessNode) and sdict[x] is None
+        ]
+        for node in second_state.nodes():
+            if isinstance(node, nodes.NestedSDFG):
+                # update parent information
+                node.sdfg.parent = first_state
+            #we only want to add the node once!    
+            try:
+                first_state.add_node(node)
```

Comment:
Why is the try / except necessary here, when is a node being added that already exists?

Reply:
You are correct - it is not. I was not able to create a case where this is needed. Likely another bug that was fixed, or SC tiredness issue